### DISCUSSION TITLE
preprocessing and crop or pad

### DISCUSSION THREAD

**Discussion by EvelyneCalista** [TorchIO-project/torchio#1129] [General]
Hi ,
I would like to ask several questions about the pre-processing transform.
1. I use transform=[tio.RescaleIntensity()] then i am intended to save the image into nifti files using "subject.image.save('*.nii'). However after saving, the Image header for the "Origin X,Y,Z" is changed example from -105 to 0. It looks like the image get flipped upside down.
 I tried with other image to check, but it looks normal (the header not get changed).
 I use torchio version0.18.63. is this a bug or did i make a mistake?
3. I would like to use croporpad, after croporpad from original size (512x512x30) to (256x256x30) the voxel spacing size now is twice larger than before. is that normal?

Thank you!

**Comment by fepegar** [comment]
Hi, @EvelyneCalista. Neither of those are expected. Can you please share some code so we can reproduce the issue?

**Comment by EvelyneCalista** [comment]
Hi @fepegar ,
thank you for your response. Please see below, also I found that with this code below only the "mask_img" will be transformed but not the "image". Please help your advice, thank you.

```python
class ResampleXY:
    def __init__(self, spacing_x, spacing_y):
        self.spacing_x = spacing_x
        self.spacing_y = spacing_y
    
    def __call__(self, subject):
        sx,sy,sz=subject.spacing
        resampling = tio.Resample((self.spacing_x, self.spacing_y, sz))
        resmpled_img = resampling(subject)
        return resmpled_img

if __name__ == "__main__":
    for id in subjid:
        save_dir = os.path.join(savepath, str(id))
        if not os.path.exists(save_dir):
            os.makedirs(save_dir)
        scan = os.path.join(rootpath,str(id)+'/scan.nii')    
        mask_label = os.path.join(rootpath,str(id)+'/label.nii')
        mask_lb = glob.glob(os.path.join(root_mask,str(id),'*.nii'))[0]
       
        crop_shape=(512,512,44)
        resize_ds_rate = (2,2,1)

        
        subject = tio.Subject(
            image=tio.ScalarImage(scan),
            mask_img = tio.ScalarImage(mask_lb),
            label = tio.LabelMap(mask_label))

        transforms = [
                    tio.ToCanonical(),                
                    ResampleXY(spacing_x=0.45, spacing_y=0.45),
                    tio.CropOrPad(crop_shape,mask_name='label'),
                    tio.RescaleIntensity()
                ]

        transform = tio.Compose(transforms)
        preprocessed = transform(subject)
        preprocessed.image.save(os.path.join(save_dir,'croppad.nii.gz'))
        preprocessed.mask_img.save(os.path.join(save_dir,'mask.nii.gz'))
```

**Reply by fepegar** [reply]
Thanks for sharing this code. Here are some thoughts:

> the Image header for the "Origin X,Y,Z" is changed example from -105 to 0

When you crop or pad an image, you change its field of view in the physical space, so this is expected. Basically, you're probably changing the position of the voxel at $(0, 0, 0)$, which is the "origin".

> It looks like the image get flipped upside down

What viewer are you using? FSL? `ToCanonical` rearranges the voxels so that the last direction changes from feet to head (inferior to superior), so maybe your images had the opposite arrangement before (i.e., its "orientation" ended with `I`) and you're using a visualization tool that is not aware of spatial metadata.

> I tried with other image to check, but it looks normal (the header not get changed)

Maybe your other image had a different header to start with?

> I use torchio version0.18.63

Please upgrade, this is very old and we're constantly adding features and fixing bugs.

> the voxel spacing size now is twice larger than before

This is the effect of your resampling operation. Your images probably had spacing $1 \times 1 \times 1$ mm and you're doubling the spacing along the first two dimensions.

---

I hope those made sense. If they still don't, please upgrade your version and share a _fully_ reproducible example demonstrating what's unexpected.

**Comment by EvelyneCalista** [comment]
Thank you for your advice, I have updated the version now.

I use ITK-SNAP, would this visualization tool good enough?
Yes, the image orientation is RPI, shall I change it into RPS? or do you think in that case I don't need to add `ToCanonical` into the preprocessing?

**Reply by fepegar** [reply]
Yes, ITK-SNAP should be aware of the spatial metadata. You don't need to change the orientation manually beforehand.
